File: ai/models_config.py
"""Centralized model configuration loaded from models.json."""

import json
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from langchain_openai import ChatOpenAI


@dataclass
class Provider:
    """Provider configuration."""

    name: str
    base_url: str | None
    api_key_env: str
    headers: dict[str, str] | None = None

    @property
    def api_key(self) -> str:
        """Get API key from environment."""
        return os.getenv(self.api_key_env, "")


@dataclass
class Model:
    """Model configuration."""

    name: str
    provider: str
    model_id: str
    extra_body: dict[str, Any] | None = None

    def get_provider(self) -> Provider:
        """Get the provider for this model."""
        return get_provider(self.provider)

    @property
    def base_url(self) -> str | None:
        """Get base URL from provider."""
        return self.get_provider().base_url

    @property
    def api_key(self) -> str:
        """Get API key from provider."""
        return self.get_provider().api_key

    @property
    def headers(self) -> dict[str, str] | None:
        """Get default headers from provider."""
        return self.get_provider().headers

    @property
    def display_name(self) -> str:
        """Display name with provider prefix."""
        return f"{self.provider}/{self.name}"


# Load config from models.json
_config_path = Path(__file__).parent.parent / "models.json"
with open(_config_path) as f:
    _config = json.load(f)

# Parse providers
_providers: dict[str, Provider] = {}
for name, data in _config["providers"].items():
    _providers[name] = Provider(
        name=name,
        base_url=data.get("base_url"),
        api_key_env=data["api_key_env"],
        headers=data.get("headers"),
    )

# Parse models
_llm_models: list[Model] = [
    Model(name=m["name"], provider=m["provider"], model_id=m["model_id"], extra_body=m.get("extra_body"))
    for m in _config["llm"]
]

_asr_models: list[Model] = [
    Model(name=m["name"], provider=m["provider"], model_id=m["model_id"]) for m in _config["asr"]
]

_tts_models: list[Model] = [
    Model(name=m["name"], provider=m["provider"], model_id=m["model_id"]) for m in _config["tts"]
]


def get_provider(name: str) -> Provider:
    """Get provider by name."""
    if name not in _providers:
        raise ValueError(f"Unknown provider: {name}")
    return _providers[name]


def get_llm_models() -> list[Model]:
    """Get all LLM models."""
    return _llm_models


def get_asr_models() -> list[Model]:
    """Get all ASR models."""
    return _asr_models


def get_tts_models() -> list[Model]:
    """Get all TTS models."""
    return _tts_models


def get_llm_by_index(index: int) -> Model:
    """Get LLM model by index, wrapping around."""
    return _llm_models[index % len(_llm_models)]


def get_default_llm() -> Model:
    """Get default LLM model (first in list)."""
    return _llm_models[0]


def get_llm_by_model_id(model_id: str) -> Model | None:
    """Find LLM model by model_id."""
    for model in _llm_models:
        if model.model_id == model_id:
            return model
    return None


def create_chat_llm(
    model_id: str,
    *,
    streaming: bool = True,
    callbacks: list[Any] | None = None,
) -> ChatOpenAI:
    """Create a ChatOpenAI client from model_id.

    Single source of truth for ChatOpenAI construction — handles
    provider headers, extra_body, and all provider-specific config.

    Raises:
        ValueError: If model_id is not found in models.json
    """
    model = get_llm_by_model_id(model_id)
    if not model:
        raise ValueError(f"Unknown model_id: {model_id!r}. Check models.json configuration.")
    kwargs: dict[str, Any] = {}
    if model.extra_body:
        kwargs["model_kwargs"] = {"extra_body": model.extra_body}
    if callbacks:
        kwargs["callbacks"] = callbacks
    return ChatOpenAI(
        base_url=model.base_url,  # pyright: ignore[reportArgumentType]
        api_key=model.api_key,  # pyright: ignore[reportArgumentType]
        model=model.model_id,  # pyright: ignore[reportArgumentType]
        streaming=streaming,
        default_headers=model.headers or {},
        **kwargs,
    )

File: desktop/lib/models/models_config.dart
/// Centralized model configuration loaded from models.json asset.
library;

import 'dart:convert';

import 'package:flutter/services.dart' show rootBundle;

/// Provider configuration.
class Provider {
  final String name;
  final String? baseUrl;
  final String apiKeyEnv;

  const Provider({
    required this.name,
    this.baseUrl,
    required this.apiKeyEnv,
  });

  factory Provider.fromJson(String name, Map<String, dynamic> json) {
    return Provider(
      name: name,
      baseUrl: json['base_url'] as String?,
      apiKeyEnv: json['api_key_env'] as String,
    );
  }
}

/// Model configuration.
class Model {
  final String name;
  final String provider;
  final String modelId;

  const Model({
    required this.name,
    required this.provider,
    required this.modelId,
  });

  factory Model.fromJson(Map<String, dynamic> json) {
    return Model(
      name: json['name'] as String,
      provider: json['provider'] as String,
      modelId: json['model_id'] as String,
    );
  }

  /// Display name with provider prefix.
  String get displayName => '$provider/$name';
}

/// Singleton holding loaded model configuration.
class ModelsConfig {
  static ModelsConfig? _instance;

  final Map<String, Provider> providers;
  final List<Model> llmModels;
  final List<Model> asrModels;
  final List<Model> ttsModels;

  ModelsConfig._({
    required this.providers,
    required this.llmModels,
    required this.asrModels,
    required this.ttsModels,
  });

  /// Load configuration from asset. Call once at app startup.
  static Future<void> load() async {
    if (_instance != null) return;

    final jsonString = await rootBundle.loadString('assets/models.json');
    final json = jsonDecode(jsonString) as Map<String, dynamic>;

    // Parse providers
    final providersJson = json['providers'] as Map<String, dynamic>;
    final providers = <String, Provider>{};
    for (final entry in providersJson.entries) {
      providers[entry.key] = Provider.fromJson(
        entry.key,
        entry.value as Map<String, dynamic>,
      );
    }

    // Parse models
    final llmJson = json['llm'] as List<dynamic>;
    final llmModels = llmJson
        .map((m) => Model.fromJson(m as Map<String, dynamic>))
        .toList();

    final asrJson = json['asr'] as List<dynamic>;
    final asrModels = asrJson
        .map((m) => Model.fromJson(m as Map<String, dynamic>))
        .toList();

    final ttsJson = json['tts'] as List<dynamic>;
    final ttsModels = ttsJson
        .map((m) => Model.fromJson(m as Map<String, dynamic>))
        .toList();

    _instance = ModelsConfig._(
      providers: providers,
      llmModels: llmModels,
      asrModels: asrModels,
      ttsModels: ttsModels,
    );
  }

  /// Get the loaded instance. Throws if not loaded.
  static ModelsConfig get instance {
    if (_instance == null) {
      throw StateError('ModelsConfig not loaded. Call ModelsConfig.load() first.');
    }
    return _instance!;
  }

  /// Get provider by name.
  Provider getProvider(String name) {
    final provider = providers[name];
    if (provider == null) {
      throw ArgumentError('Unknown provider: $name');
    }
    return provider;
  }

  /// Get LLM model by index, wrapping around.
  Model getLlmByIndex(int index) {
    return llmModels[index % llmModels.length];
  }

  /// Get default LLM model (first in list).
  Model get defaultLlm => llmModels.first;

  /// Get default ASR model (first in list).
  Model get defaultAsr => asrModels.first;

  /// Get default TTS model (first in list).
  Model get defaultTts => ttsModels.first;
}
